sync/src/relayer/block_transactions_process.rs

```diff
@@ -59,24 +82,67 @@ impl<'a> BlockTransactionsProcess<'a> {
                     self.peer
                 );
 
-                let transactions: Vec<Transaction> =
-                    FlatbuffersVectorIterator::new(cast!(self.message.transactions())?)
-                        .map(TryInto::try_into)
-                        .collect::<Result<_, FailureError>>()?;
+                let txs_len = transactions.len();
 
                 BlockTransactionsVerifier::verify(&compact_block, &indexes, &transactions)?;
 
                 let ret = self.relayer.reconstruct_block(compact_block, transactions);
 
-                // TODO Add this (compact_block, peer) into RecentRejects if reconstruct_block failed?
-                // TODO Add this block into RecentRejects if accept_block failed?
-                if let Ok(block) = ret {
-                    pending.remove();
-                    self.relayer
-                        .accept_block(self.nc.as_ref(), self.peer, block);
-                    return Ok(Status::Accept);
+                match ret {
+                    Ok(block) => {
+                        if block.header().transactions_root()
+                            != compact_block.header.transactions_root()
+                        {
+                            if compact_block.short_ids.len() == txs_len {
+                                self.relayer.shared().insert_block_status(
```

Comment:
We can never assert the block is invalid only when the merkle root does not match.

Reply:
a mistake. the code is inconsistent with the comment.